compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala

```diff
@@ -248,6 +248,138 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
+     * `this` value before the super constructor call, which is invalid.
+     * While most of the time this analysis is only an optimization, in those
+     * (rare) situations it is required for correctness. See for example
+     * the partest `run/t8733.scala`.
+     */
+
+    private var statifyCandidateMethodsThatReferenceThis: scala.collection.Set[Symbol] = null
+
+    /** Is that given method a statify-candidate?
+     *
+     *  If a method is a statify-candidate, we will analyze whether it actually
+     *  needs its `this` value. If it does not, we will compile it as a static
+     *  method in the IR.
+     *
+     *  A method is a statify-candidate if it is a lifted method (a local def)
+     *  or the target of an anonymous function.
+     *
+     *  TODO Currently we also require that the method owner not be a JS type.
+     *  We should relax that restriction in the future.
+     */
+    private def isStatifyCandidate(sym: Symbol): Boolean =
+      (sym.isLiftedMethod || sym.isDelambdafyTarget) && !isJSType(sym.owner)
+
+    /** Do we compile the given method as a static method in the IR?
+     *
+     *  This is true if one of the following is true:
+     *
+     *  - It is `isStaticMember`, or
+     *  - It is a statify-candidate method and it does not reference `this`.
+     */
+    private def compileAsStaticMethod(sym: Symbol): Boolean = {
+      sym.isStaticMember || {
+        isStatifyCandidate(sym) &&
+        !statifyCandidateMethodsThatReferenceThis.contains(sym)
+      }
+    }
+
+    /** Finds all statify-candidate methods that reference `this`, inspired by Delambdafy. */
+    private final class ThisReferringMethodsTraverser extends Traverser {
+      // the set of statify-candidate methods that directly refer to `this`
+      private val roots = mutable.Set.empty[Symbol]
+
+      // for each statify-candidate method `m`, the set of statify-candidate methods that call `m`
+      private val methodReferences = mutable.Map.empty[Symbol, mutable.Set[Symbol]]
+
+      def methodReferencesThisIn(tree: Tree): collection.Set[Symbol] = {
+        traverse(tree)
+        propagateReferences()
+      }
+
+      private def addRoot(symbol: Symbol): Unit =
+        roots += symbol
+
+      private def addReference(from: Symbol, to: Symbol): Unit =
+        methodReferences.getOrElseUpdate(to, mutable.Set.empty) += from
+
+      private def propagateReferences(): collection.Set[Symbol] = {
+        val result = mutable.Set.empty[Symbol]
+
+        def rec(symbol: Symbol): Unit = {
+          // mark `symbol` as referring to `this`
+          if (result.add(symbol)) {
+            // `symbol` was not yet in the set; propagate further
+            for {
+              referrers <- methodReferences.get(symbol)
+              referrer <- referrers
+            } {
+              // referrer calls `symbol`, so it transitively references `this`
+              rec(referrer)
+            }
```

Comment:
```suggestion
            methodReferences.getOrElse(symbol, Nil).foreach(rec(_))
```

For readability? (IMO the nested comment is not necessary).